sdlib/d/gc/global.d

```diff
@@ -46,14 +53,42 @@ private:
 		roots = (cast(const(void*)[]*) ptr)[0 .. length];
 
 		import d.gc.range;
-		roots[index] = makeRange(range);
+		if (range.length == 0) {
+			roots[index] = cast(void*[]) range;
+		} else {
+			roots[index] = makeRange(range);
+		}
+	}
+
+	void removeRootsImpl(const void* ptr) {
+		assert(mutex.isHeld(), "Mutex not held!");
+
+		// Search in reverse, since it's most likely for things to be removed
+		// in the reverse order they were added.
+		foreach_reverse (i; 0 .. roots.length) {
+			if (cast(void*) roots[i].ptr == ptr) {
+				auto length = roots.length - 1;
+				roots[i] = roots[length];
+				import d.gc.tcache;
+				auto newRoots = threadCache
+					.realloc(roots.ptr, length * void*[].sizeof, true);
+				roots = (cast(const(void*)[]*) newRoots)[0 .. length];
+				return;
+			}
+		}
 	}
 
 	void scanRootsImpl(ScanDg scan) {
 		assert(mutex.isHeld(), "Mutex not held!");
 
 		foreach (range; roots) {
-			scan(range);
+			// Adding a range of length 0 is like pinning the given range
```

Comment:
Use /** style comment for multi line comments.

Reply:
I was following precedent, but OK.